services/galley/src/Galley/API/Update.hs

```diff
@@ -1055,6 +1057,86 @@ addMembersUnqualified lusr zcon cnv (Invite users role) = do
   let qusers = fmap (tUntagged . qualifyAs lusr) (toNonEmpty users)
   addMembers lusr zcon (tUntagged (qualifyAs lusr cnv)) (InviteQualified qusers role)
 
+-- | Replace conversation members by computing the difference between desired and
+-- current members, then executing removals followed by additions within a commit
+-- lock.
+replaceMembers ::
+  forall r.
+  ( Member BackendNotificationQueueAccess r,
+    Member BrigAPIAccess r,
+    Member ConversationStore r,
+    Member (Error InternalError) r,
+    Member (ErrorS ('ActionDenied 'AddConversationMember)) r,
+    Member (ErrorS ('ActionDenied 'LeaveConversation)) r,
+    Member (ErrorS ('ActionDenied 'RemoveConversationMember)) r,
+    Member (ErrorS 'ConvAccessDenied) r,
+    Member (ErrorS 'ConvNotFound) r,
+    Member (ErrorS 'InvalidOperation) r,
+    Member (ErrorS 'NotConnected) r,
+    Member (ErrorS 'NotATeamMember) r,
+    Member (ErrorS 'TooManyMembers) r,
+    Member (ErrorS 'MissingLegalholdConsent) r,
+    Member (ErrorS 'GroupIdVersionNotSupported) r,
+    Member (Error FederationError) r,
+    Member (Error NonFederatingBackends) r,
+    Member (Error UnreachableBackends) r,
+    Member ExternalAccess r,
+    Member FederatorAccess r,
+    Member NotificationSubsystem r,
+    Member (Input Env) r,
+    Member (Input Opts) r,
+    Member Now r,
+    Member LegalHoldStore r,
+    Member ProposalStore r,
+    Member Random r,
+    Member TeamStore r,
+    Member TinyLog r,
+    Member TeamCollaboratorsSubsystem r,
+    Member E.MLSCommitLockStore r
+  ) =>
+  Local UserId ->
+  ConnId ->
+  Qualified ConvId ->
+  InviteQualified ->
+  Sem r ()
+replaceMembers lusr zcon qcnv (InviteQualified desiredUsers role) = do
+  lcnv <- ensureLocal lusr qcnv
+  conv <- getConversationWithError lcnv
+
+  -- Check team permissions for desired members
+  when (null conv.metadata.cnvmParent) $
+    mapErrorS @OperationDenied @('ActionDenied 'AddConversationMember) $
+      forM_ conv.metadata.cnvmTeam $ \tid -> do
+        forM_ desiredUsers $ \u -> do
+          mTeamMembership <- E.getTeamMember tid $ qUnqualified u
+          forM_ (mTeamMembership >>= permissionsRole . Wire.API.Team.Member.getPermissions) $
+            permissionCheck JoinRegularConversations . Just
+
+  -- Get current members (excluding the requesting user)
+  let currentMembers = Set.fromList $ map (\m -> Qualified m.id_ (tDomain lcnv)) (toList conv.localMembers)
+      desiredMembersSet = Set.fromList $ toList desiredUsers
+      toRemove = Set.difference currentMembers desiredMembersSet
+      toAdd = Set.difference desiredMembersSet currentMembers
+
+  -- If both sets are empty, return Unchanged
+  unless (Set.null toRemove && Set.null toAdd) $ do
+    -- Execute additions and removals sequentially within commit lock
```

Comment:
I'm confused. Where's the commit lock?

Reply:
My bad, outdated comment, removed, thanks.